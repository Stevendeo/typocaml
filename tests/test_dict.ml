(**************************************************************************)
(*                                                                        *)
(*  Copyright (c) 2024 OCamlPro                                           *)
(*                                                                        *)
(*  All rights reserved.                                                  *)
(*  This file is distributed under the terms of the                       *)
(*  OCamlPro license.                                                     *)
(*                                                                        *)
(**************************************************************************)

(* let () = Typocaml.Log.activate_debug () *)

let (++) m str = Typocaml.add (String.lowercase_ascii str) () m

let pp_unit ppf () = Fmt.pf ppf "x"

let pp t =
  Typocaml.pp ~pp_data:pp_unit Format.std_formatter t

let pp_l ppf l =
  List.iter
    (fun (word, (), typos) -> Fmt.pf ppf "%s(%i);@." word typos)
    l

let new_pp () =
  Fmt.pf Format.std_formatter "@.--@."

let pp_opt pp ppf = function
  | None -> Fmt.pf ppf "<none>"
  | Some v -> pp ppf v

let dict =
  let text = "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium totam rem aperiam eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit sed quia consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt Neque porro quisquam est qui dolorem ipsum quia dolor sit amet consectetur adipisci velit sed quia non numquam eius modi tempora incidunt ut labore et dolore magnam aliquam quaerat voluptatem Ut enim ad minima veniam quis nostrum exercitationem ullam corporis suscipit laboriosam nisi ut aliquid ex ea commodi consequatur Quis autem vel eum iure reprehenderit qui in ea voluptate velit esse quam nihil molestiae consequatur vel illum qui dolorem eum fugiat quo voluptas nulla pariatur" in
  let l = String.split_on_char ' ' text in
  List.fold_left (++) Typocaml.empty l

let autocomplete_i =
  Typocaml.get_autocompletions "i" dict

let autocomplete_ip =
  Typocaml.get_autocompletions "ip" dict

let autocomplete_ip' =
  Typocaml.get_autocompletions "p" autocomplete_i

let ok1 = Typocaml.find_opt ~max_typo:1 "voluptatam" dict
let ko1 = Typocaml.find_opt ~max_typo:1 "nuwwa" dict
let ok2 = Typocaml.find_opt ~max_typo:2 "nuwwa" dict

let ok3 = Typocaml.find_opt ~max_typo:1 "nulal" dict

let ca1 = Typocaml.autocomplete_autocorrect ~max_typo:0 "ca" dict
let ca2 = Typocaml.autocomplete_autocorrect ~max_typo:1 "ca" dict
let co1 = Typocaml.autocomplete_autocorrect ~max_typo:0 "co" dict

let () =
  pp dict;
  new_pp ();
  pp autocomplete_i;
  new_pp ();
  pp autocomplete_ip;
  new_pp ();
  pp autocomplete_ip';
  new_pp ();
  (* The two autocompletions should be equal *)
  Fmt.pf Format.std_formatter "%b" (autocomplete_ip = autocomplete_ip');
  new_pp ();
  (* Finding words in the tree. *)
  Fmt.pf Format.std_formatter "voluptatam(1): %a@." (pp_opt pp_unit) ok1;
  new_pp ();
  (* This one should not be found, it has two mistakes. *)
  Fmt.pf Format.std_formatter "nuwwa(1): %a@." (pp_opt pp_unit) ko1;
  new_pp ();
  (* With two typos, we can find it. *)
  Fmt.pf Format.std_formatter "nuwwa(2): %a@." (pp_opt pp_unit) ok2;
  new_pp ();
  Fmt.pf Format.std_formatter "nulal(1): %a@." (pp_opt pp_unit) ok3;
  new_pp ();
  (* Return all autocompletions of ca with 0 & 1 typos *)
  Fmt.pf Format.std_formatter "ca(0):[@.%a]@." pp_l ca1;
  Fmt.pf Format.std_formatter "ca(1):[@.%a]@." pp_l ca2;
  Fmt.pf Format.std_formatter "co(0):[@.%a]@." pp_l co1;
