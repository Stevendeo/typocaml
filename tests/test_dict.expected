| a ->
  | b -> x
  | c ->
    | c ->
      | u ->
        | s ->
          | a ->
            | n ->
              | t ->
                | i ->
                  | u ->
                    | m -> x
  | d -> x
    | i ->
      | p ->
        | i ->
          | s ->
            | c ->
              | i -> x
  | l ->
    | i ->
      | q ->
        | u ->
          | a ->
            | m -> x
          | i ->
            | d -> x
  | m ->
    | e ->
      | t -> x
  | p ->
    | e ->
      | r ->
        | i ->
          | a ->
            | m -> x
  | r ->
    | c ->
      | h ->
        | i ->
          | t ->
            | e ->
              | c ->
                | t ->
                  | o -> x
  | s ->
    | p ->
      | e ->
        | r ->
          | n ->
            | a ->
              | t ->
                | u ->
                  | r -> x
  | u ->
    | t -> x
      | e ->
        | m -> x
| b ->
  | e ->
    | a ->
      | t ->
        | a ->
          | e -> x
| c ->
  | o ->
    | m ->
      | m ->
        | o ->
          | d ->
            | i -> x
    | n ->
      | s ->
        | e ->
          | c ->
            | t ->
              | e ->
                | t ->
                  | u ->
                    | r -> x
          | q ->
            | u ->
              | a ->
                | t ->
                  | u ->
                    | r -> x
              | u ->
                | n ->
                  | t ->
                    | u ->
                      | r -> x
    | r ->
      | p ->
        | o ->
          | r ->
            | i ->
              | s -> x
| d ->
  | i ->
    | c ->
      | t ->
        | a -> x
  | o ->
    | l ->
      | o ->
        | r -> x
          | e -> x
            | m -> x
              | q ->
                | u ->
                  | e -> x
            | s -> x
| e ->
  | a -> x
    | q ->
      | u ->
        | e -> x
  | i ->
    | u ->
      | s -> x
  | n ->
    | i ->
      | m -> x
  | o ->
    | s -> x
  | r ->
    | r ->
      | o ->
        | r -> x
  | s ->
    | s ->
      | e -> x
    | t -> x
  | t -> x
  | u ->
    | m -> x
  | x -> x
    | e ->
      | r ->
        | c ->
          | i ->
            | t ->
              | a ->
                | t ->
                  | i ->
                    | o ->
                      | n ->
                        | e ->
                          | m -> x
    | p ->
      | l ->
        | i ->
          | c ->
            | a ->
              | b ->
                | o -> x
| f ->
  | u ->
    | g ->
      | i ->
        | a ->
          | t -> x
        | t -> x
| i ->
  | l ->
    | l ->
      | o -> x
      | u ->
        | m -> x
  | n -> x
    | c ->
      | i ->
        | d ->
          | u ->
            | n ->
              | t -> x
    | v ->
      | e ->
        | n ->
          | t ->
            | o ->
              | r ->
                | e -> x
  | p ->
    | s ->
      | a -> x
        | m -> x
      | u ->
        | m -> x
  | s ->
    | t ->
      | e -> x
  | u ->
    | r ->
      | e -> x
| l ->
  | a ->
    | b ->
      | o ->
        | r ->
          | e -> x
          | i ->
            | o ->
              | s ->
                | a ->
                  | m -> x
    | u ->
      | d ->
        | a ->
          | n ->
            | t ->
              | i ->
                | u ->
                  | m -> x
| m ->
  | a ->
    | g ->
      | n ->
        | a ->
          | m -> x
        | i -> x
  | i ->
    | n ->
      | i ->
        | m ->
          | a -> x
  | o ->
    | d ->
      | i -> x
    | l ->
      | e ->
        | s ->
          | t ->
            | i ->
              | a ->
                | e -> x
| n ->
  | a ->
    | t ->
      | u ->
        | s -> x
  | e ->
    | m ->
      | o -> x
    | q ->
      | u ->
        | e -> x
    | s ->
      | c ->
        | i ->
          | u ->
            | n ->
              | t -> x
  | i ->
    | h ->
      | i ->
        | l -> x
    | s ->
      | i -> x
  | o ->
    | n -> x
    | s ->
      | t ->
        | r ->
          | u ->
            | m -> x
  | u ->
    | l ->
      | l ->
        | a -> x
    | m ->
      | q ->
        | u ->
          | a ->
            | m -> x
| o ->
  | d ->
    | i ->
      | t -> x
  | m ->
    | n ->
      | i ->
        | s -> x
| p ->
  | a ->
    | r ->
      | i ->
        | a ->
          | t ->
            | u ->
              | r -> x
  | e ->
    | r ->
      | s ->
        | p ->
          | i ->
            | c ->
              | i ->
                | a ->
                  | t ->
                    | i ->
                      | s -> x
  | o ->
    | r ->
      | r ->
        | o -> x
| q ->
  | u ->
    | a ->
      | e -> x
        | r ->
          | a ->
            | t -> x
      | m -> x
      | s ->
        | i -> x
    | i -> x
      | a -> x
      | s -> x
        | q ->
          | u ->
            | a ->
              | m -> x
    | o -> x
| r ->
  | a ->
    | t ->
      | i ->
        | o ->
          | n ->
            | e -> x
  | e ->
    | m -> x
    | p ->
      | r ->
        | e ->
          | h ->
            | e ->
              | n ->
                | d ->
                  | e ->
                    | r ->
                      | i ->
                        | t -> x
| s ->
  | e ->
    | d -> x
    | q ->
      | u ->
        | i -> x
  | i ->
    | t -> x
  | u ->
    | n ->
      | t -> x
    | s ->
      | c ->
        | i ->
          | p ->
            | i ->
              | t -> x
| t ->
  | e ->
    | m ->
      | p ->
        | o ->
          | r ->
            | a -> x
  | o ->
    | t ->
      | a ->
        | m -> x
| u ->
  | l ->
    | l ->
      | a ->
        | m -> x
  | n ->
    | d ->
      | e -> x
  | t -> x
| v ->
  | e ->
    | l -> x
      | i ->
        | t -> x
    | n ->
      | i ->
        | a ->
          | m -> x
    | r ->
      | i ->
        | t ->
          | a ->
            | t ->
              | i ->
                | s -> x
  | i ->
    | t ->
      | a ->
        | e -> x
  | o ->
    | l ->
      | u ->
        | p ->
          | t ->
            | a ->
              | s -> x
              | t ->
                | e -> x
                  | m -> x
--
| l ->
  | l ->
    | o -> x
    | u ->
      | m -> x
| n -> x
  | c ->
    | i ->
      | d ->
        | u ->
          | n ->
            | t -> x
  | v ->
    | e ->
      | n ->
        | t ->
          | o ->
            | r ->
              | e -> x
| p ->
  | s ->
    | a -> x
      | m -> x
    | u ->
      | m -> x
| s ->
  | t ->
    | e -> x
| u ->
  | r ->
    | e -> x
--
| s ->
  | a -> x
    | m -> x
  | u ->
    | m -> x
--
| s ->
  | a -> x
    | m -> x
  | u ->
    | m -> x
--
true
--
voluptatam(1): x

--
nuwwa(1): <none>

--
nuwwa(2): x

--
nulal(1): x

--
ca(0):[
]
ca(1):[
corporis(1);
consequuntur(1);
consequatur(1);
consectetur(1);
commodi(1);
autem(1);
aut(1);
aspernatur(1);
architecto(1);
aperiam(1);
amet(1);
aliquid(1);
aliquam(1);
adipisci(1);
ad(1);
accusantium(1);
ab(1);
eaque(1);
laudantium(1);
laboriosam(1);
labore(1);
magni(1);
magnam(1);
natus(1);
pariatur(1);
ratione(1);
]
co(0):[
corporis(0);
consequuntur(0);
consequatur(0);
consectetur(0);
commodi(0);
]
